### PR TITLE
Introduce `JsonIndexAccessor`

### PR DESCRIPTION
**Pull request by sbrannen**
## Overview

This PR contains two commits.

The first polishes `JsonPropertyAccessor` and associated tests.

The second introduces a `JsonIndexAccessor` as a complement to the existing `JsonPropertyAccessor`.

When a `JsonIndexAccessor` is registered with the SpEL `EvaluationContext` alongside a registered `JsonPropertyAccessor`, JSON arrays can be consistently indexed via integer literals (e.g., `[1]`) instead of string literals representing integers (e.g., `['1']`).

## Related Issues

- #9383

### CODE REVIEW

**Review by artembilan — CHANGES_REQUESTED** [review]
Hi @sbrannen !

Thank you very much for the contribution.

I'd like to have a couple clarifications:

1. You say in your commit message ` instead of string literals representing integers (e.g., ['1']).`. But that's not fully true according to our existing tests, e.g. `JsonPropertyAccessorTests.testArrayLookup()` or the next one.
I might miss something and this sentence is for another use-case.
2. Would you mind to give me some explanation for docs to justify this new `JsonIndexAccessor` and what benefits it has when it is used in combination with `JsonPropertyAccessor`? 
Or feel free to make a respective change into `spel.adoc`. 😄

**Comment by sbrannen** [comment]
Hi @artembilan,

> Thank you very much for the contribution.

You're very welcome. 😎 

> You say in your commit message `instead of string literals representing integers (e.g., ['1']).`. But that's not fully true according to our existing tests, e.g. `JsonPropertyAccessorTests.testArrayLookup()` or the next one.
> I might miss something and this sentence is for another use-case.

On `main`, `JsonPropertyAccessorTests.testArrayLookup()` explicitly **wraps** the `ArrayNode`, which means that it is no longer an `ArrayNode`. Instead, it becomes an `ArrayNodeAsList` which is a `java.util.List`, and you can always index into a `List` with integer indexes.

In my PR, I renamed that method to `arrayLookupWithIntegerIndexAndExplicitWrapping()` to make it clear that explicit, manual wrapping is occurring in the test method.

Whereas, users would likely not manually wrap an `ArrayNode` that they make available in a SpEL expression, meaning that users would have to supply `['1']` instead of `[1]` as the expression.

Without the `JsonIndexAccessor`, the first index must always be a `String` if you attempt to index directly into an `ArrayNode`. 

I made this clear in the new `nestedArrayLookupWithStringIndexAndThenIntegerIndex()` test method, which works with or without the `JsonIndexAccessor`; whereas, **all** of the tests in `JsonPropertyAccessorTests.ArrayNodeTests` assert that integer indexing is not supported with only a `JsonPropertyAccessor`. On the other hand, all of the tests in `JsonIndexAccessorTests.ArrayNodeTests` pass with integer indexes.

Does that clarify things for you?

**Comment by sbrannen** [comment]
> Would you mind to give me some explanation for docs to justify this new `JsonIndexAccessor` and what benefits it has when it is used in combination with `JsonPropertyAccessor`?

To be honest, I originally thought that introducing a `JsonIndexAccessor` would provide greater benefit to Spring Integration (SI) users; however, after thoroughly diving into the current feature set in SI, I realized that the _wrapping_ functionality in `JsonPropertyAccessor` already covers a lot of ground, and I was able to reuse that wrapping functionality in `JsonIndexAccessor` (indirectly via `JsonPropertyAccessor.typedValue()`).

For example, as mentioned in my previous comment, the "wrapping" of an `ArrayNode` in a `List` allows for integer indexing into the list.

And that wrapping occurs automatically whenever a property is accessed on a `JsonNode` which returns an `ArrayNode` or whenever indexing into an `ArrayNodeAsList` returns an `ArrayNode` wrapped in an `ArrayNodeAsList`.

The only time an `ArrayNode` is not automatically wrapped in an `ArrayNodeAsList` is when the root context object is an `ArrayNode` or when a property or method invocation returns an `ArrayNode` directly.

In summary, the primary (perhaps only) benefit of `JsonIndexAccessor` is that it allows JSON arrays to be consistently indexed via integer literals, with the key word being "consistently" (meaning "always").

Having said that, I still think it's beneficial to have the `JsonIndexAccessor` since it allows people to write idiomatic expressions like `[1][2]` instead of `['1']['2']`.

> Or feel free to make a respective change into `spel.adoc`. 😄

Oh... you want me to write documentation, too? 🤣

Kidding aside, I'll see if I find time to look at / edit the `spel.adoc`.

**Comment by sbrannen** [comment]
> 2\. Would you mind to give me some explanation for docs to justify this new `JsonIndexAccessor` and what benefits it has when it is used in combination with `JsonPropertyAccessor`?
>         Or feel free to make a respective change into `spel.adoc`.

I took a look at the [Spring Expression Language (SpEL)](https://docs.spring.io/spring-integration/reference/6.4-SNAPSHOT/spel.html) section of the reference manual, and based on the status quo, I don't think it makes sense to add documentation for the `JsonIndexAccessor` _yet_.

Rationale: you'll first need to decide if you want to introduce first-class configuration support for `IndexAccessor`s in general.

For example, you can add a `PropertyAccessor` to `IntegrationEvaluationContextFactoryBean` and `IntegrationSimpleEvaluationContextFactoryBean` (which is not currently documented in the reference manual), via the `<spel-property-accessors>` XML namespace element, and via a `spelPropertyAccessorRegistrar` `@Bean` method. But... there's currently no easy way for SI users to register an `IndexAccessor` (either a custom one of their own or the `JsonIndexAccessor`.

**Comment by sbrannen** [comment]
> Rationale: you'll first need to decide if you want to introduce first-class configuration support for `IndexAccessor`s in general.

I created https://github.com/spring-projects/spring-integration/issues/9436, so that that issue can be addressed separately, independent of the fate of this PR.

**Review by artembilan — APPROVED** [review]
OK, Sam.
Nothing to complain about.
Merging as is.
I'll take care about `IndexAccessor` injection support and respective docs with the issue you have created.

Thank you for contribution!

**Comment by sbrannen** [comment]
> Merging as is. 

Awesome!

> I'll take care about `IndexAccessor` injection support and respective docs with the issue you have created.

Sounds good. I'll keep an eye on the status of that issue.

> Thank you for contribution!

🙇
